### ISSUE TITLE
Wrong type inference in result of classmethod when call it from subclass

### ISSUE THREAD

**Issue by qmmp123** [microsoft/pyright#6530]
**Describe the bug**
Wrong type inference in result of classmethod when call it from subclass

**Code or Screenshots**
```python
from typing import Generic, TypeVar


T = TypeVar("T")


class Foo(Generic[T]):

    def __init__(self, container: T):
        self.container = container
    
    @classmethod
    def foo(cls, container: T) -> "Foo[T]":
        return cls(container)


class Bar(Foo):
    pass

foo = Foo.foo([1,2,3,4])
foo.container  # here pyright shows list[int] type
bar = Bar.foo([1,2,3,4])
bar.container  # here pyright shows Uknown type
bar  # here pyright shows Foo[Uknown] type
```

**VS Code extension or command-line**
pyright 1.1.337

**Comment by erictraut** [comment]
Pyright's behavior here is correct. The class `Bar` derives from `Foo`. The `Foo` class is generic and takes a single type argument, but you haven't provided any here. When no type argument is provided, type checkers will assume `Any`. Pyright uses the name `Unknown` to represent an "implicit" `Any`. 

If your intent is for `Bar` to also be generic, you need to provide a type parameter for `Bar`, like this:

```python
class Bar(Foo[T]):
    pass
```
